README

Updated for version 3.8.7

For information about building the FreeForm data handler, see the
INSTALL file.

This file provides basic information about the FreeForm handler and its
configuration. For more detailed information, see the Server
Installation Guide at http://www.opendap.org. Also note that there is
also a special manual for the FreeForm server that covers writing
format description files in detail available at the OPeNDAP web site.

For general information about configuration of the base software for
the OPeNDAP server, see the README that comes with that server and/or
the Server Installation Guide that can be found at www.opendap.org.

Test data are also installed, so after installing this handler, Hyrax
will have data to serve providing an easy way to test your new
installation and to see how a working bes.conf should look. To use
this, make sure that you first install the bes, and that dap-server
gets installed too.  Finally, every time you install or reinstall
handlers, make sure to restart the BES and OLFS.

--------------------------------------------------------------------------

freeform_handler is a data handler which uses National Geophysical Data
Center's FreeForm/FFND API to serve data. To use with the Hyrax, you'll
need BES 3.9.0 or greater.

The handler can read binary, ASCII, and dBASE data files using
FreeForm/FFND format descriptions. It can return both array and
sequence data types. The data in the data file is read then served
using the appropriate type according to the format description in the
required accompanying format file.

All the data and associated blocks (header information) are processed.
Basic information about the input file is given as Global Attributes as
is any header information. The file's data variables are mapped into
the closest DAP variable type using the FreeForm/FFND external format
file.

Customizations for Remote Sensing Systems (RSS) Data files

Two parameters in the configuration file control features that enable
using data from RSS. The parameter FF.RSSFormatSupport is used to 
activate parsing the RSS-format-specific .fmt files. The parameter
FF.RSSFormatFiles is used to specify a directory other than the data 
where the .fmt (and optional ancillary .das files) will be found. If 
this is not given or is the empty string, the handler will look in the
same directory as the data. If it is given, the handler will look _only_
in the specified directory.

FF.RSSFormatSupport = yes
FF.RSSFormatFiles = /usr/local/RSS


WRITING FORMAT DESCRIPTION FILES

Any data file served must be accompanied by a format description file
(<file_name>.fmt). The format description file defines the formats of
headers and data in the data file. The following shows a few examples
of data description files and their translation into DAP variables, for
more information about format descriptions, please see the
OPeNDAP-FreeFrom's user guide.

1) Format to read a dBASE sequence file 
   
<code>
       dbase_input_data "ISEE3_data"
       t 1 15 double 6
       Bx 16 30 double 6
       By 31 45 double 6
       Bz 46 60 double 6
       Bx2 61 75 double 6
</code>

   The above format file will map into the following DAP data set:

<code>
     Dataset {
       Sequence {
        Float64 t;
        Float64 Bx;
        Float64 By;
        Float64 Bz;
        Float64 Bx2;
       } ISEE3_data;
     } test5;
</code>


2) Format to read a binary array file

<code>
   binary_input_data "array_in"
   x 1 4 ARRAY["lat" 1 to 10]["lon" 1 to 10] of int32 0
   y 5 12 ARRAY["records" 1 to 10] of float64 15
</code>

   The above format file will map into the following DAP data set:

<code>
   Dataset {
      Int32 x[lat = 10][lon = 10];
      Float64 y[records = 10];
   } test3;
</code>


3) Format to read an ASCII sequence with a file header

<code>
      ASCII_file_header "header"
      count 1 2 short 0
      description 4 24 char 0

      ASCII_input_data "data"
      value 1 2 short 0
</code>

   The above format file will map into the following DAP data set:

<code>
       Dataset {
	  Sequence {
	  Int32 value;
	  } data;
	} test0;
</code>

4) The FreeForm library contains a useful data processing/transformation 
library. Here's an example of how to tweak data written on a big-endian
machine so it can be served from a little-endian one - include the following
at the start of the format file:

<code>
input_eqv begin constant
    data_byte_order text big_endian
end constant 
</code>

Thanks to Denis Nadeau for this information.

For more examples see the test files in the ff-test subdirectory. There is
a manual that describes FreeForm and handler on our web page - opendap.org -
in the User Documentation section.      

--------------------------------------------------------------------------

Note: The FreeForm library, which is in the subdirectory 'FFND,' carries
the following notice regarding its copyright protection:

<code>
 * CAVEAT:
 * No claims are made as to the suitability of the accompanying
 * source code for any purpose.  Although this source code has been
 * used by the NOAA, no warranty, expressed or implied, is made by
 * NOAA or the United States Government as to the accuracy and
 * functioning of this source code, nor shall the fact of distribution
 * constitute any such endorsement, and no responsibility is assumed
 * by NOAA in connection therewith.  The source code contained
 * within was developed by an agency of the U.S. Government.
 * NOAA's National Geophysical Data Center has no objection to the
 * use of this source code for any purpose since it is not subject to
 * copyright protection in the U.S.  If this source code is incorporated
 * into other software, a statement identifying this source code may be
 * required under 17 U.S.C. 403 to appear with any copyright notice.
</code>



NEWS
News for version 3.9.5

Updates since 2016-05-10, see the ChangeLog

News for version 3.9.4

Updates since 2016-01-26, see the ChangeLog

News for version 3.9.3

Minor updates, see the ChangeLog

News for version 3.9.2

Minor updates, see the ChangeLog

News for version 3.9.1

Minor updates, see the ChangeLog

Version 3.8.8

Minor Updates

Version 3.8.7

Fixed the behavior of BaseType::read() to match libdap 3.12.1

Fixed some of the date() server functions WRT the clang 5 compiler on OSX.

Version 3.8.6

Build fixes.

Version 3.8.5

Test updates.

Version 3.8.3

Build fixes

Version 3.8.2 (14 Sept 2010)

Build fix for the freeform tools.

Version 3.8.1 (5 May 2010)

The handler now implements its own transfer_attributes code to merge
attributes from its DAS response into the DDS.

There are also handful of build improvements.

version 3.8.0 (2 Feb 2010)

Removed Dejagnu tests and added autotest tests. 
Made changes to allow for a Shrew build project and added 'cccc' targets
so that a metric snapshot can be taken.
 
version 3.7.12 (9 June 2009)

Minor fix - fixed broken dejagnu tests, soon to be replaced with autotest
tests. 

Version 3.7.11 (16 March 2009)

Constraint fixes.

Support for the service interface.

version 3.7.10 (29 Jan 2009)

Bug fixes and updates for the new libdap and bes code.

Version 3.7.9 (26 June 2008)

Minor enhancements and configuration upgrades

Version 3.7.8 (29 Feb 2008)

Updates for bes 3.6.0.

Configuration updates for pkg and rpm builds, libdap and bes checking, and
pkg configuration.

Version 3.7.7 (14 Nov 2007)

pkg and rpm build fixes

Version 3.7.6 (14 Nov 2007)

Don't edit bes.conf during make install, but document the bes-ff-data.sh
script and the bes-conf make target in README.

Version 3.7.5 (13 March 2007)

Added automatic configuration of the handler using 'make install.' Now the
handler is automatically configured in the bes.conf file in the default
location. Test data are also installed, so after installing this handler
Hyrax will have data to serve providing an easy way to test your new install
and to see how the a working bes.conf should look. To use this, make sure
that you first install the bes, and that dap-server gets intalled too.
Finally, every time you install or reinstall handlers, make sure to restart
the BES and OLFS.

Version 3.7.4

Improved error handling.

Comes bundled with test data.

Build fixes.

Version 3.7.3

Updates for the 1.4 release of Server 4. Also included are bug fixes, memory
leak fixes, and performance enhencements.

Version 3.7.2

Now builds as a libtool module.

The module name is no longer hardcoded.

Version 3.7.1

Fixed a bug (#435) where functions in different handlers with the same names
clashed when the handlers were loaded as shared object libraries. 

Patrick fixed a problem with the help text.

Version 3.7.0

This version of the FreeForm handler includes an executable for use with
Server3 and a shared-object library (aka module) for use with Server4. This
is a source release to support the Server4 1.2 beta release and has not been
tested with Server3, although there should be no problems.

Version 3.6.1

Fixes for Mac OS/X:

1. A compile-time issue was resolved. With Tiger, configure did not set the
compile-time symbol FREEFORM to the correct value. This caused the compile to
fail.

2. Since the Mac is a big-endian machine (well, excpet for the new intel
macs...) some of the binary data tests failed. I've marked those as expected
to fail so that the whole suite does not fail beacuse of this.

Note that we are not making binary packages for OS/X, so this will be a source-
only release. These changes will not affect the Linux distributions. The test
changes may affect the Solaris dist, but not in an operational way.

Version 3.6.0

Updated rpm spec file using a patch from Patrice Dumas.

Updated the code so that it compiles with libdap 3.6.0.

Version 3.5.0

This version of the FreeForm format handler is essentially the same as the
previous version (3.4) with the exception that it now builds like the other
handlers and works with the dap-server 3.5 base software.


FFByte.cc

// -*- mode: c++; c-basic-offset:4 -*-

// This file is part of ff_handler a FreeForm API handler for the OPeNDAP
// DAP2 data server.

// Copyright (c) 2005 OPeNDAP, Inc.
// Author: James Gallagher <jgallagher@opendap.org>
//
// This is free software; you can redistribute it and/or modify it under the
// terms of the GNU Lesser General Public License as published by the Free
// Software Foundation; either version 2.1 of the License, or (at your
// option) any later version.
//
// This software is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
// or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
// License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
//
// You can contact OPeNDAP, Inc. at PO Box 112, Saunderstown, RI. 02874-0112.

// (c) COPYRIGHT URI/MIT 1997-99
// Please read the full copyright statement in the file COPYRIGHT.
//
// Authors: reza (Reza Nekovei)


// FreeFrom sub-class implementation for FFByte,...FFGrid.
// The files are patterned after the subcalssing examples
// Test<type>.c,h files.
//
// ReZa 6/18/97

#include "config_ff.h"

static char rcsid[] not_used = {"$Id$"};


#include <string>

#include "FFByte.h"
#include "util_ff.h"
#include "util.h"

extern long BufPtr;
extern char * BufVal;

// This `helper functions' creates a pointer to the a FFByte and returns
// that pointer. It takes the same arguments as the class's ctor. If any of
// the variable classes are subclassed (e.g., to make a new Byte like
// HDFByte) then the corresponding function here, and in the other class
// definition files, needs to be changed so that it creates an instance of
// the new (sub)class. Continuing the earlier example, that would mean that
// NewByte() would return a HDFByte, not a Byte.
//
// It is important that these function's names and return types do not change
// - they are called by the parser code (for the dds, at least) so if their
// names changes, that will break.
//
// The declarations for these functions (in util.h) should *not* need
// changing.

FFByte::FFByte(const string &n, const string &d) : Byte(n, d)
{
}

BaseType *
FFByte::ptr_duplicate()
{
    return new FFByte(*this);
}

bool
FFByte::read()
{
    if (read_p()) // nothing to do
	return false;

    if (BufVal){ // Data in cache
	char * ptr = BufVal+BufPtr;
	val2buf((void *) ptr);
	set_read_p(true);

	BufPtr += width();
	return false;
    }
    else {

	return false;
    }
}

// $Log: FFByte.cc,v $
// Revision 1.10  2003/02/10 23:01:52  jimg
// Merged with 3.2.5
//
// Revision 1.9.2.1  2002/12/18 23:30:42  pwest
// gcc3.2 compile corrections, mainly regarding the using statement
//
// Revision 1.9  2000/10/11 19:37:56  jimg
// Moved the CVS log entries to the end of files.
// Changed the definition of the read method to match the dap library.
// Added exception handling.
// Added exceptions to the read methods.
//
// Revision 1.8  1999/05/27 17:02:22  jimg
// Merge with alpha-3-0-0
//
// Revision 1.7.2.1  1999/05/20 21:39:27  edavis
// Fix spelling of COPYRIGHT and remove some #if 0 stuff.
//
// Revision 1.7  1999/05/04 02:55:36  jimg
// Merge with no-gnu
//
// Revision 1.6  1999/03/26 20:03:31  jimg
// Added support for the Int16, UInt16 and Float32 datatypes
//
// Revision 1.5.12.1  1999/05/01 04:40:29  brent
// converted old String.h to the new std C++ <string> code
//
// Revision 1.5  1998/08/13 20:24:21  jimg
// Fixed read mfunc semantics
//
// Revision 1.4  1998/08/12 21:20:51  jimg
// Massive changes from Reza. Compatible with the new FFND library
//
// Revision 1.3  1998/04/21 17:13:43  jimg
// Fixes for warnings, etc
//
// Revision 1.2  1998/04/16 18:11:00  jimg
// Sequence support added by Reza
